acnportal/acnsim/tests/test_simulator.py

```diff
@@ -0,0 +1,157 @@
+        np.testing.assert_allclose(self.simulator.pilot_signals,
+            np.zeros((len(self.simulator.network.station_ids), len(self.simulator.event_queue))))
+
+    def test_correct_on_init_charging_rates(self):
+        np.testing.assert_allclose(self.simulator.charging_rates,
+            np.zeros((len(self.simulator.network.station_ids), len(self.simulator.event_queue))))
+
+    def test_update_schedules_unequal_lengths(self):
+        new_schedule = {'PS-001' : [24], 'PS-002' : [16, 24]}
+        with self.assertRaises(InvalidScheduleError):
+            self.simulator._update_schedules(new_schedule)
+
+    def test_update_schedules_not_in_network(self):
+        new_schedule = {'PS-001' : [24, 16], 'PS-004' : [16, 24]}
+        with self.assertRaises(KeyError):
+            self.simulator._update_schedules(new_schedule)
+
+    def test_update_schedules_valid_schedule(self):
+        new_schedule = {'PS-001' : [24, 16], 'PS-002' : [16, 24]}
+        self.simulator._update_schedules(new_schedule)
+        np.testing.assert_allclose(self.simulator.pilot_signals[:, :2], np.array([[24, 16], [16, 24], [0, 0]]))
+
+    def test_run(self):
+        # Integration test
+        # -- Experiment Parameters ---------------------------------------------------------------------------------------------
+        timezone = pytz.timezone('America/Los_Angeles')
+        start = timezone.localize(datetime(2018, 9, 5))
+        end = timezone.localize(datetime(2018, 9, 6))
+        period = 5  # minute
+        voltage = 220  # volts
+        default_battery_power = 32 * voltage / 1000 # kW
+        site = 'caltech'
+
+        # -- Network -----------------------------------------------------------------------------------------------------------
+        cn = sites.caltech_acn(basic_evse=True, voltage=voltage)
+
+        # -- Events ------------------------------------------------------------------------------------------------------------
+        API_KEY = 'DEMO_TOKEN'
+        events = acndata_events.generate_events(API_KEY, site, start, end, period, voltage, default_battery_power)
+
+
+        # -- Scheduling Algorithm ----------------------------------------------------------------------------------------------
+        sch = EarliestDeadlineFirstAlgo(increment=1)
+
+        # -- Simulator ---------------------------------------------------------------------------------------------------------
+        self.sim = Simulator(deepcopy(cn), sch, deepcopy(events), start, period=period, max_recomp=1, verbose=False)
+        self.sim.run()
+
+        edf_algo_true_info_dict = {}
+        edf_algo_new_info_dict = {}
+        info_fields = ['pilot_signals', 'charging_rates', 'peak', 'ev_history', 'event_history']
+        work_dir = os.path.join(os.path.dirname(__file__), "edf_algo_true_info_fields")
+        for field in info_fields:
+            with open(os.path.join(work_dir, "edf_algo_true_"+field+".p"), 'rb') as info_file:
```

Comment:
Perhaps use human readable files such as csv or json. Pickle files are dangerous since they can execute malicious code when unpicked.

Reply:
We'll use json for this, but we should consider writing methods to serialize the entire simulator.

Reply:
Agreed. We should add whole simulator serialization. See #18